de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoice/model/validator/C_Invoice.java

```diff
@@ -254,4 +257,29 @@ private void deleteInvoiceLines(final I_C_Invoice invoice)
 			InterfaceWrapperHelper.delete(line);
 		}
 	}
+
+	@DocValidate(timings = { ModelValidator.TIMING_BEFORE_COMPLETE })
+	public void linkInvoiceToPaymentIfNeeded(final I_C_Invoice invoice)
+	{
+		final I_C_Order order = invoice.getC_Order();
+		if (order != null && Services.get(IDocTypeBL.class).isPrepay(order.getC_DocType()) &&  order.getC_Payment_ID() > 0)
+		{
+			final I_C_Payment payment = order.getC_Payment();
+			payment.setC_Invoice_ID(invoice.getC_Invoice_ID());
+			InterfaceWrapperHelper.save(payment);
+
+			Services.get(IAllocationBL.class).autoAllocateSpecificPayment(invoice, payment, true);
+		}
+	}
+
+	@DocValidate(timings = { ModelValidator.TIMING_BEFORE_COMPLETE })
+	public void allocateInvoiceAgainstPaymentIfNeeded(final I_C_Invoice invoice)
+	{
+		final I_C_Order order = invoice.getC_Order();
```

Comment:
note sure I get it right, but this method seems to be the same as `linkInvoiceToPaymentIfNeeded()` minus the `payment.setC_Invoice_ID(...);` part..why do we need it?

Reply:
Yes, right! wrong moment; the second should be TIMING_AFTER_COMPLETE